File: app/src/test/java/org/mifos/mobilebanking/RegistrationPresenterTest.java
package org.mifos.mobilebanking;

import android.content.Context;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mifos.mobilebanking.api.DataManager;
import org.mifos.mobilebanking.models.register.RegisterPayload;
import org.mifos.mobilebanking.presenters.RegistrationPresenter;
import org.mifos.mobilebanking.ui.views.RegistrationView;
import org.mifos.mobilebanking.util.RxSchedulersOverrideRule;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import io.reactivex.Observable;
import okhttp3.ResponseBody;

import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * Created by Chirag Gupta on 11/29/17.
 */
@RunWith(MockitoJUnitRunner.class)
public class RegistrationPresenterTest {

    @Rule
    public final RxSchedulersOverrideRule mOverrideSchedulersRule = new RxSchedulersOverrideRule();

    @Mock
    Context context;

    @Mock
    DataManager dataManager;

    @Mock
    RegistrationView view;

    @Mock
    ResponseBody responseBody;

    private RegistrationPresenter presenter;
    private RegisterPayload registerPayload;

    @Before
    public void setUp() throws Exception {
        presenter = new RegistrationPresenter(dataManager, context);
        presenter.attachView(view);

        registerPayload = FakeRemoteDataSource.getRegisterPayload();
    }

    @After
    public void tearDown() throws Exception {
        presenter.detachView();
    }

    @Test
    public void testRegisterUser() throws Exception {
        when(dataManager.registerUser(registerPayload)).thenReturn(Observable.just(responseBody));

        presenter.registerUser(registerPayload);
        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view).showRegisteredSuccessfully();
        verify(view, never()).showError("");
    }

    @Test
    public void testRegisterUserFails() throws Exception {
        when(dataManager.registerUser(registerPayload)).thenReturn(Observable.
                <ResponseBody>error(new RuntimeException()));

        presenter.registerUser(registerPayload);
        verify(view).showProgress();
        verify(view).hideProgress();
        verify(view, never()).showRegisteredSuccessfully();
        verify(view).showError("");
    }
}

File: app/src/test/java/org/mifos/mobile/RegistrationPresenterTest.kt
package org.mifos.mobile

import android.content.Context

import io.reactivex.Observable

import okhttp3.ResponseBody

import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

import org.mifos.mobile.api.DataManager
import org.mifos.mobile.models.register.RegisterPayload
import org.mifos.mobile.presenters.RegistrationPresenter
import org.mifos.mobile.ui.views.RegistrationView
import org.mifos.mobile.util.RxSchedulersOverrideRule

import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnitRunner

/**
 * Created by Chirag Gupta on 11/29/17.
 */
@RunWith(MockitoJUnitRunner::class)
class RegistrationPresenterTest {
    @Rule
    @JvmField
    val mOverrideSchedulersRule = RxSchedulersOverrideRule()

    @Mock
    var context: Context? = null

    @Mock
    var dataManager: DataManager? = null

    @Mock
    var view: RegistrationView? = null

    @Mock
    var responseBody: ResponseBody? = null
    private var presenter: RegistrationPresenter? = null
    private var registerPayload: RegisterPayload? = null

    @Before
    @Throws(Exception::class)
    fun setUp() {
        presenter = RegistrationPresenter(dataManager!!, context)
        presenter?.attachView(view!!)
        registerPayload = FakeRemoteDataSource.registerPayload
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        presenter?.detachView()
    }

    @Test
    @Throws(Exception::class)
    fun testRegisterUser() {
        Mockito.`when`<Observable<ResponseBody?>?>(dataManager?.registerUser(registerPayload)).thenReturn(Observable.just(responseBody))
        presenter?.registerUser(registerPayload)
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view)?.showRegisteredSuccessfully()
        Mockito.verify(view, Mockito.never())?.showError("")
    }

    @Test
    @Throws(Exception::class)
    fun testRegisterUserFails() {
        Mockito.`when`(dataManager?.registerUser(registerPayload)).thenReturn(Observable.error(RuntimeException()))
        presenter?.registerUser(registerPayload)
        Mockito.verify(view)?.showProgress()
        Mockito.verify(view)?.hideProgress()
        Mockito.verify(view, Mockito.never())?.showRegisteredSuccessfully()
        Mockito.verify(view)?.showError("")
    }
}